[IconAnnotator] - Add IconAnnotator to Mark Objects with Custom Icons/Images

dfreilich opened the pull request:
# Description

This PR adds an IconAnnotator class, which can annotate images. For best results, the icon should be a transparent icon without a background, so that it will come up clearly against any images. 

![zidane-icon-annotated](https://github.com/roboflow/supervision/assets/7035673/9787f365-fec1-44d9-92a4-b4e5a5d09c2a)

Some open questions for discussion:
1. The demo image for the annotation didn't follow the standard placement for the icon (it seems to be above the image, rather than at the center top border of the detection). I added logic to center the icon horizontally, but I wasn't sure about adding that vertically as well. 
2. To what degree is it important to have color adaption for the icon annotation? It can look very nice, but it also introduces more error cases, for icons that aren't transparent, and will just show up as a colored box as a result. 

I'd love to hear your thoughts on it.

A Collab notebook which demonstrates the change is [here](https://colab.research.google.com/drive/1ridnlRDB89RrWznzaKZuGzSppzgPmySf?usp=sharing)

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [x] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

This change includes a unit test (the first created for annotations) to verify that the annotation works, and the result closely matches a so-called golden file

## Docs

-   [ ] Docs updated? What were the changes:

Closes #460

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/roboflow/supervision?pullRequest=930) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/roboflow/supervision?pullRequest=930) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/roboflow/supervision?pullRequest=930) it.</sub>

SkalskiP commented:
Hi @dfreilich 👋🏻 ! Thank you very much for your interest in Supervision.

Given that we don't yet have a concrete idea for end-to-end testing of annotators, I wouldn't want to introduce it as part of this PR. I kindly ask you to remove the following files from this PR:
- `test/data/zidane.jpg`
- `test/data/zidane-icon-annotated.jpg`
- `test/data/icons8-diamond-50.png`
- `test/annotators/test_core.py`

SkalskiP commented on supervision/annotators/core.py:
```diff
@@ -1763,3 +1763,113 @@ def validate_custom_values(
+                int(xy[detection_idx, 1]),
+            )
+
+            # Takes the alpha channel from the original icon
```
I think we can remove most of those comments and replace them with more descriptive variable names.

dfreilich commented on supervision/annotators/core.py:
```diff
@@ -1763,3 +1763,113 @@ def validate_custom_values(
+        self.position = position
+        icon = cv2.imread(icon_path, cv2.IMREAD_UNCHANGED)
+        if icon is None:
+            print(f"Error: Couldn't load the icon image from {icon_path}")
```
Makes sense! I raised a `FileNotFoundError`, which seems to address the error in the clearest way, but I can change it to another exception if you'd like

dfreilich commented on supervision/annotators/core.py:
```diff
@@ -1763,3 +1763,113 @@ def validate_custom_values(
+
+            # The current positions of the anchors don't account for annotations that
+            # have mass. This visually centers the anchor, given the size of the icon
+            x_offset = 0
```
It was most glaring in the center case, which is why I handled it. I can also handle it in left/right cases as well, to make sure that it's outside of the bounds or on the boundary itself. `y_offset` we definitely can add (albeit we'll need to be careful to make sure it doesn't run into conflict with boundaries of the image), but I was waiting to get feedback from the project on whether `x_offset` was desired first before adding it

dfreilich commented:
@SkalskiP Thanks for the review! I addressed the comments, and removed the unit tests I had created. With regards to the changes in [acceptance.md](https://github.com/roboflow/supervision/blob/develop/docs/annotators.md), I assumed that it didn't make sense to add the `IconAnnotator` to the carousel at the top (like we do here: https://github.com/roboflow/supervision/blob/8599345a3c00c26e7a05d7c81e23bf1d5cf8b8e2/docs/annotators.md?plain=1#L8-L27) until we have an agreed upon example that's hosted on `media.roboflow.com`, like the rest of the examples, but I can add one if you'd like.

LinasKo commented:
I've modernized and cleaned up this one, but will shift focus for now. I'll be back after the metrics are complete.
@dfreilich, if you have the time and energy, feel free to jump in, and we can include it in the next release 😉 

I believe one change is needed - instead of specifying an icon at the start, we could take a `Union[str, List[str]]` in `annotate`, and either load the image once, or display different images per-object. We'll need to change the code from padding-unpadding to checking image boundaries, however, as we wouldn't know the size in advance. OR we could read all images and check the max size.

Here's my tests: [Colab](https://colab.research.google.com/drive/1b6J4GXghSWkhEKMDByu-VFSKjb9UyZ6q?usp=sharing)

Steps afterwards:
- [ ] Create an image to display in the [model carousel](https://supervision.roboflow.com/latest/detection/annotators/)
- [ ] Check if the docs are rendered correctly

LinasKo commented:
Hey @onuralpszr,

Got a second to look at the code for this one, by any chance? 😉
